src/ray/raylet/scheduling_resources.h

```diff
@@ -119,6 +119,37 @@ class ResourceSet {
   std::unordered_map<std::string, double> resource_capacity_;
 };
 
+using ResourceIdSet = std::unordered_map<std::string, std::vector<std::pair<int64_t, double>>>;
+
+/// \class FractionalResourceIDSet
+/// \brief Encapsulates and operates on a set of resources, including CPUs,
+/// GPUs, and custom labels. For a given resource, this keeps track of IDs for
+/// that resource and fractional quantities.
+class FractionalResourceIDSet {
+ public:
+  /// \brief Constructs FractionalResourceIDSet from the specified resource map.
+  FractionalResourceIDSet(const ResourceSet &resource_set);
+
+  /// \brief Empty FractionalResourceIDSet destructor.
+  ~FractionalResourceIDSet();
+
+  bool Contains(const ResourceSet &resource_set);
+
+  ResourceIdSet Acquire(const ResourceSet &resource_set);
+
+  void ReturnResources(const ResourceIdSet &resource_id_set);
+
+  const std::string ToString() const;
+
+ private:
+  // The available resources. For each resource name, there is a vector of IDs of
+  // the whole resource IDs that are available and a vector of fractional resources,
+  // where each element of the pair, is an ID and the amount of that ID that is available.
+  std::unordered_map<std::string,
+                     std::pair<std::vector<int64_t>,
+                               std::vector<std::pair<int64_t, double>>>> resource_availabilities_;
```

Comment:
The structure of the value is too complex, it is better to create a struct for it. It is easy to understand the every element by the variable name.